### ISSUE TITLE
A rule for checking CSS variables in shorthand properties 

### ISSUE THREAD

**Issue by kachkaev** [stylelint/stylelint#6911]
### What is the problem you're trying to solve?

I spent the whole working day investigating an issue with [`react-to-print`](https://github.com/gregnb/react-to-print) which turned out to be pretty deep and dark glitch in the CSSOM spec itself!

- https://github.com/w3c/csswg-drafts/issues/2515

When JavaScript relies on [`styleElement.sheet.cssRules[i].cssText`](https://github.com/gregnb/react-to-print/blob/de4c04b5472ccf00e3fdcd8f333fb4faf066e0b2/src/index.tsx#L451-L454) to copy the existing CSS, some of the valid rules ‘get lost in translation’. If we define a shorthand property like `padding`, `margin`, `background` etc. alongside its longhands (`padding-left`, `background-repeat` etc.), the value of the shorthand property disappears from `cssText` if it refers to a CSS variable 🤯

```scss
padding: 42px; // ok
padding-left: var(--some-other-var);
// ↓
padding-top: 42px;
padding-right: 42px;
padding-bottom: 42px;
padding-left: var(--some-other-var);

padding: var(--some-var); // is lost
padding-left: var(--some-other-var);
// ↓
padding-top:
padding-right:
padding-bottom:
padding-left: var(--some-other-var);
```

This limitation is very unfortunate, but it kinda makes sense. Imagine `--some-var` being equal to `"42px 10px 42px 0px"` – what should happen then?

The same unexpected behavior is shared between all three browser engines (Chromium, Safari, FF). So it’s not an implementation bug, but rather a limitation of a spec.

Third-party reproduction: https://jsfiddle.net/9gft8hkr/3/

Own reproduction: https://codesandbox.io/s/focused-wood-lh725v

Similar complaint in Chromium bug tracker:
https://bugs.chromium.org/p/chromium/issues/detail?id=1218159#c12

The only workaround for this situation that I have found is replacing

```scss
padding: var(--some-var); // to be lost
padding-left: var(--some-other-var);
```

with

```scss
padding-top: var(--some-var);
padding-right: var(--some-var);
padding-bottom: var(--some-var);
padding-left: var(--some-other-var);
```

This is only needed when the app relies on `styleElement.sheet.cssRules[i].cssText`, which can be pretty deeply nested in project dependencies.

### What solution would you like to see?

The issue I have discovered must be rather rare because it only happens at the intersection of:

- usage of CSS vars
- usage of any tooling that relies on `styleElement.sheet.cssRules[i].cssText` API
- presence of shorthand CSS property next to one or more longhands _in a single CSS rule_

Whoever gets into this unfortunate trap may get stuck for ages, so it would be nice if Stylelint could provide some guardrails in a form of a new rule. I have scanned our codebase and have manually removed shorthands in a couple of cases, but there is not guarantee that I have fixed all of them nor that new bugs are not going to be introduced.

Not sure how to name the new potential rule, but it should get triggered when a longhand CSS property is added after a shorthand property that uses `var`. I am ready to help with more feedback but I am relatively new to Stylelint, so happy for the community to decide on the best way forward!

Cheers 🙌

**Comment by ybiquitous** [comment]
@kachkaev Thanks for sharing your experience and proposal with the detailed reproduction!

This behavior may cause a headache in some situations, but I'd like to know if we should add a built-in rule because the behavior affects only CSSOM, not actual CSS.

This new rule might be redundant for people who want to keep CSS code simple.

So, at this point, it seems to make sense to me to provide this rule as a [Stylelint plugin](https://stylelint.io/developer-guide/plugins). Although, I'd like to hear others' opinions.

---

BTW, I confirmed the behavior on my Chrome Dev Tool.

In CSSOM, `padding-` values except for `padding-bottom` are lost,

<img width="234" alt="image" src="https://github.com/stylelint/stylelint/assets/473530/5f154a07-f235-49af-807f-7c944764770a">

on the other hand, in computed styles, the `padding-` values are still valid. Weird 😅 

<img width="231" alt="image" src="https://github.com/stylelint/stylelint/assets/473530/381fb54e-6e10-4016-b13f-f1a63e5b6c5e">

**Comment by jeddy3** [comment]
Closing as best suited as a [plugin](https://stylelint.io/developer-guide/plugins). If you do create it, please add it to Aweseome Stylelint so that others can find it.

**Comment by kachkaev** [comment]
Makes sense! I don’t have capacity to create a plugin soon, but it’s good that we have this knew knowledge stored in the issues. When I was investigating a bug, it took me a while to find the right info!
